booking-deeplink-specification.md

```diff
@@ -0,0 +1,24 @@
+
+## Functional overview
+
+The booking flow with a deep link starts when a passenger clicks on a journey search result in the MaaS app. This search result MUST open the deep link contained in the [DeepLink](https://github.com/klaxit/open-carpooling-service-specification/blob/c34cbfdea9991bfd3453d0d0822066411dcfa219/carpool-api-en.yml#L268) parameter returned by the [CarpoolJourney](https://github.com/klaxit/open-carpooling-service-specification/blob/c34cbfdea9991bfd3453d0d0822066411dcfa219/carpool-api-en.yml#L140) object in the [carpoolJourneys](https://github.com/klaxit/open-carpooling-service-specification/blob/master/carpool-api-en.yml#L10) response. This parameter MUST be provided by the carpool operator responsible for their implementation of the search API public specification. It contains the full URL and its corresponding parameters needed for the mobile device to redirect the passenger from the app displaying the search results to the operator's app.
+
+One obstacle to redirecting the passenger is whether they have the operator's app already installed in their mobile device or not. If not, the deep link URL SHOULD be capable of redirecting the passenger to the corresponding store instead. At the operator's discretion, this URL can also temporarily store the URL parameters for later retrieval. There are third-party services as well that can provide such a solution. This way, after the app is installed, the parameters can be recovered to automatically redirect the passenger from the onboarding to the booking flow as if the app had always been installed.
+
+If the app was already installed and the passenger was already registered with the operator service, then clicking on the deep link will automatically and seamlessly redirect the passenger to the operator's booking flow. It is up to the operator to decide where to start this flow and what to do with the parameters sent to the MaaS app in the search to provide the passenger with the best booking experience. It is up to the passenger to decide whether to book the clicked journey or not.
+
+If the passenger decides to not to book the journey being displayed (or it is not anymore available), the operator SHOULD provide the means for the passenger to go back to the search results. It is then up to the passenger to decide whether to select another result and repeat the above process. If they decide to book it, it is at the operator's discretion to decide how to continue guiding the passenger through the booking process. Once the booking is confirmed, the operator MUST send the booking details back to the MaaS if the following conditions are met. One, the passenger's account on the operator platform is linked to the MaaS using the [Connect specification](https://github.com/klaxit/open-carpooling-service-specification/blob/master/connect-specification.md). Two, the MaaS must provide an implementation of the booking API specification (tbd) for the operator to consume.
```

Comment:
We don't provide some kind of "back button" that would go back to the browser / MaaS app. From my tests on BBCD, you don't either. I don't think the flow after the offer is displayed should be specified (left to operator's decision).
Also, I think "Once the booking is confirmed [...]" should not be included in the "deeplink" specification. Both  should be used to offer a complete booking experience but it should also be possible to use them separately. Maybe renaming the current document in `standard-covoiturage.md` and splitting the two aspects in "Booking" > "Deep linking" and "Booking" > "Usage feedback" would make the spec more modular.

Reply:
>  I don't think the flow after the offer is displayed should be specified (left to operator's decision).

I propose to replace the "SHOULD" with "MAY" here, to make it more explicit that this is strictly left to the operator's discretion. To me it still feels worth mentioning as a best practice recommended UX. 

> Also, I think "Once the booking is confirmed [...]" should not be included in the "deeplink" specification.

It's not a deeplink specification but a booking flow specification _with_ a deep link. This part is anyway an optional step if the Connect part is not integrated. Not sure it is worth a separate spec just for it.

Reply:
> I propose to replace the "SHOULD" with "MAY" here, to make it more explicit that this is strictly left to the operator's discretion. To me it still feels worth mentioning it as a recommended UX.

It would mean that to date, neither of us is implementing the  "recommended UX". A bit difficult to agree to that...

> It's not a deeplink specification but a booking flow specification with a deep link. This part is anyway an optional step if the Connect part is not integrated. Not sure it is worth a separate spec just for it.

To me, it must be separated at least in a different section (/module) so a MaaS app can choose to implement one ([a] the deeplink) without the other ([b] send booking back to the MaaS app). The fact that the pre-requisites are not the same ([b] require the "connect") and that to date, all operators implement [a] but not [b] clearly campaigns for me to a clear separation between both.

Reply:
> To me, it must be separated at least in a different section (/module)

Still not totally clear to me what you are trying to achieve. Could you please make a separate PR on top of mine here with what you have in mind?

Reply:
To me, it seams important in the case of "MaaS to operators" use cases to have a back button to get back to the MaaS. The user arrives on the operator's platform just because he got the link from the MaaS. The operator being too captive after that by "preventing" to get back to the MaaS is not a right way to do for me. Having some operators providing the back button and others not, at their discretion, is even worse as depending on the journey chosen, it will be different. Then, why would an operator choose to implement that if it's better to "lock users in their app" (of course, lock is exagerated, but it's the idea) like competitors ?

SHOULD looks better to me except if there are real technical reasons preventing that (and not only "operator's choice").

Reply:
As discussed today, we don't have a technical solution to date to support this experience. The `MAY` seems definitely the way to go.

@riclage we let you update the `SHOULD` to `MAY`, reflect it in the schema and create the issue stating that we should work on technical solution allowing that in the future ?

Reply:
Done.